libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -641,176 +678,423 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
+ * further synchronize if necessary.
+ *
+ ***********************************************************************/
+
+template<_CUDA_VSTD::size_t _Align, typename _Group>
+inline __device__
+__completion_mechanism __dispatch_memcpy_async_any_to_any(_Group const & __group, char * __dest_char, char const * __src_char, _CUDA_VSTD::size_t __size, uint32_t __allowed_completions, uint64_t* __bar_handle) {
+    __cp_async_fallback<_Align>(__group, __dest_char, __src_char, __size);
+    return __completion_mechanism::__sync;
+}
+
+template<_CUDA_VSTD::size_t _Align, typename _Group>
+inline __device__
+__completion_mechanism __dispatch_memcpy_async_global_to_shared(_Group const & __group, char * __dest_char, char const * __src_char, _CUDA_VSTD::size_t __size, uint32_t __allowed_completions, uint64_t* __bar_handle) {
+    NV_IF_TARGET(NV_PROVIDES_SM_90, (
+        const bool __can_use_complete_tx = __allowed_completions & uint32_t(__completion_mechanism::__mbarrier_complete_tx);
+        _LIBCUDACXX_DEBUG_ASSERT(__can_use_complete_tx == (nullptr != __bar_handle), "Pass non-null bar_handle if and only if can_use_complete_tx.");
+        if _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 (16 <= _Align) {
+            if (__can_use_complete_tx && __isShared(__bar_handle)) {
+                __cp_async_bulk_shared_global(__group, __dest_char, __src_char, __size, __bar_handle);
+                return __completion_mechanism::__mbarrier_complete_tx;
             }
         }
-        else
-        {
-            __strided_memcpy<_Native_alignment>(__destination, __source, __size, __group.thread_rank(), __group.size());
+        // Fallthrough to SM 80..
+    ));
+
+    NV_IF_TARGET(NV_PROVIDES_SM_80, (
+        const bool __can_use_async_group = __allowed_completions & uint32_t(__completion_mechanism::__async_group);
+        if _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 (4 <= _Align) {
+            if (__can_use_async_group) {
+                __cp_async_shared_global_impl<_Align>(__group, __dest_char, __src_char, __size);
+                return __completion_mechanism::__async_group;
+            }
         }
+        // Fallthrough..
+    ));
 
-        __memcpy_arrive_on_impl::__arrive_on(__sync, __is_async);
-        , NV_ANY_TARGET,
-            __strided_memcpy<_Native_alignment>(__destination, __source, __size, __group.thread_rank(), __group.size());
-    )
+    __cp_async_fallback<_Align>(__group, __dest_char, __src_char, __size);
+    return __completion_mechanism::__sync;
+}
 
-    return __is_async;
+template<_CUDA_VSTD::size_t _Align, typename _Group>
+inline __device__
+__completion_mechanism __dispatch_memcpy_async_device(_Group const & __group, char * __dest_char, char const * __src_char, _CUDA_VSTD::size_t __size, uint32_t __allowed_completions, uint64_t* __bar_handle) {
+    // Dispatch based on direction of the copy: global to shared, shared to
+    // global, etc.
+
+    // CUDA compilers < 12.3 may not propagate assumptions about the state space
```

Comment:
Technically we should not add comments about unreleased cuda versions, so this should be 
```suggestion
    // CUDA compilers <= 12.2 may not propagate assumptions about the state space
```